### PR TITLE
69 landing page historical and more data buttons

### LINKED ISSUE #69: Landing Page - Historical and More Data Buttons
### Description

We need buttons to allow the user to view more details on the current run or allow them to view the stored historical runs for the day

### Acceptance Criteria

- [ ] More details brings you to the real time data graph
- [ ] Historical pops up a modal
- [ ] Modal allows users to select a run that is stored in the database at the time
- [ ] Shows a horizontal gallery of cards for each run as shown in the mockup
- [ ] Once a user selects a run bring it to the real time data graph but show only the data that was collected on that run
- [ ] Pass the run id to the historical graph component
- [ ] Takes in all the runs to be able to display this information

### Proposed Solution

Use material ui components for modals and such

### Mocks

Screenshot 2023-10-12 at 9 46 57 AM
Screenshot 2023-10-12 at 9 47 01 AM
Screenshot 2023-10-23 at 9 51 07 PM
Screenshot 2023-10-31 at 10 54 15 PM

### PR DESCRIPTION
**Pull request by leoleader**
## Changes

11/28 update
back from my hiatus of being sick and dying in my classes and thanksgiving break i cant believe its been 3 weeks since i updated this PR that sucks
- Caught my branch up on changes on main i missed
- Created custom button component to use for more details and historical buttons and future buttons
- Updated getAllRuns serverService call for Historical and minor updates in how routing is done for more data

11/6 update
- Updated UI in response to feedback (i thought adding title and border looked nice, let me know what u think)
- Moved carousel component into its own folder and files
- minor cleaning up code

11/1
- Added History and More Data Buttons
- More Data redirects to the real time data graph, currently just sends to empty route since there is no real time data graph component yet
- History brings up a modal with a carousel of all the runs, allowing the user to navigate through runs and select a run
- Mat UI did not have a gallery/carousel component so I researched online and found that PrimeNG carousel module was prob the best to use, so I used that, which means needs to be added to dependencies.

## Notes

TO DO STILL ->
- clean up the carousel because it's style and html are both a lil messy and rows/columns can become custom components
- have select run button actually go to runID page
- adding modal indicators (so dont have to navigate one run at a time)

## Screenshots
Screen Shot 2023-11-28 at 4 39 32 AM
Screen Shot 2023-11-28 at 4 39 51 AM

Closes #69

### CODE REVIEW

File: angular-client/src/components/history/carousel.component.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<p-carousel [value]="runs" [numVisible]="1" [numScroll]="1" [circular]="true" [showIndicators]="false">
+  <ng-template let-run pTemplate="item">
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/carousel.component.html:2]: why do you have this template thing

File: angular-client/src/components/history/carousel.component.html [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    <div class="runcard">
+      <div class="row">
+        <h3 class="left">Run #{{ run.id }}</h3>
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/carousel.component.html:5]: I think we should make a custom styled component for this and then use it where you use all the left classes

File: angular-client/src/components/history/carousel.component.html [hunk 3]
```diff
@@ -0,0 +1,21 @@
+<p-carousel [value]="runs" [numVisible]="1" [numScroll]="1" [circular]="true" [showIndicators]="false">
+  <ng-template let-run pTemplate="item">
+    <div class="runcard">
+      <div class="row">
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/carousel.component.html:4]: yeah we should make like a row layout component instead. Or use a grid, look at the grid documentation for MUI, i didnt exactly like their implementation of it compared to the react grid but its still something to consider

File: angular-client/src/components/history/carousel.component.html [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+        <h3 class="right">System: {{ run.systemName }}</h3>
+      </div>
+      <div class="row">
+        <h3 class="right">Location: {{ run.locationName }}</h3>
+      </div>
+    </div>
+  </ng-template>
+</p-carousel>
+<div class="vertical-center">
+  <button mat-button (click)="onNoClick()">Back</button>
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/carousel.component.html:19]: can you not just call mat-button?

File: angular-client/src/components/history/carousel.component.html [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  <button mat-button (click)="onNoClick()" cdkFocusInitial>Select this Run</button>
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/carousel.component.html:20]: oh yeah you need to on hover styles as well

File: angular-client/src/components/history/history.component.css [hunk 6]
```diff
@@ -0,0 +1,58 @@
+.button {
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/history.component.css:1]: im not gonna look at the styles until weve finalized the UI

File: angular-client/src/components/history/history.component.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+@Component({
+  selector: 'history',
+  templateUrl: './history.component.html',
+  styleUrls: ['./history.component.css']
+})
+export class HistoryButton {
+  runs: Run[];
+  serverService = new APIService();
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/history.component.ts:22]: we should pass the service into this component

File: angular-client/src/components/history/history.component.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+  openDialog(): void {
+    this.dialog.open(CarouselRun, {
+      width: '550px',
+      height: '200px',
+      data: { runs: this.runs }
+    });
+  }
+}
+
+@Component({
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/history.component.ts:43]: this should be in its own folder

File: angular-client/src/components/history/history.component.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    @Inject(MAT_DIALOG_DATA) public data: DialogData
+  ) {
+    this.runs = data.runs;
+  }
+
+  onNoClick(): void {
+    this.dialogRef.close();
+  }
+
+  ngOnInit() {}
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/history.component.ts:64]: why implement onInit if your not gonna use it

File: angular-client/src/components/more-details/more-details.component.html [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<router-outlet></router-outlet>
```
Comment (Peyton-McKee) [inline angular-client/src/components/more-details/more-details.component.html:1]: why do you need this?

File: angular-client/src/components/more-details/more-details.component.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'more-details',
+  templateUrl: './more-details.component.html',
+  styleUrls: ['./more-details.component.css']
+})
+export default class MoreDetails {}
```
Comment (Peyton-McKee) [inline angular-client/src/components/more-details/more-details.component.ts:8]: shouldnt your onClick functions be defined in here? not in the html itself?

File: angular-client/src/components/history/history.component.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    // this.runs = this.serverService.query<Run[]>(getAllRuns).data.subscribe.arguments;
+    this.runs = [
+      { id: 1, locationName: 'Boston', driverName: 'Bruh', systemName: 'Yuh', time: 1100 },
+      { id: 2, locationName: 'China', driverName: 'Bruh', systemName: 'Yuh', time: 1200 },
+      { id: 3, locationName: 'China', driverName: 'Bruh', systemName: 'Yuh', time: 1300 }
+    ];
+  }
+
+  openDialog(): void {
+    this.dialog.open(CarouselRun, {
```
Comment (Peyton-McKee) [inline angular-client/src/components/history/history.component.ts:35]: where is dialog defined?

**Review by Peyton-McKee — CHANGES_REQUESTED** [review]
nice, left mostly questions/. We also need to make all the changes i mentioned in slack as well.

File: angular-client/src/app/app.module.ts [hunk 13]
```diff
@@ -25,9 +29,11 @@ import LandingPage from 'src/components/landing-page/landing-page.component';
...
     MatToolbarModule,
     MatSidenavModule,
-    MatListModule
+    MatListModule,
+    MatDialogModule,
+    CarouselModule
   ],
   providers: [],
-  bootstrap: [LandingPage]
+  bootstrap: [LandingPage, MoreDetails, HistoryButton, CarouselRun]
```
[the remark this answers is not in the document]
Comment (leoleader) [reply angular-client/src/app/app.module.ts:37]: i do not, i thought i did oops

File: angular-client/src/components/history/carousel.component.html (hunk 4 above)
Comment (leoleader) [reply angular-client/src/components/history/carousel.component.html:19]: the tag <button mat-button> is the correct usage according to the docs

Comment (leoleader) [reply angular-client/src/components/history/carousel.component.html:19]: Screen Shot 2023-11-06 at 8 09 20 PM
examples

File: angular-client/src/components/more-details/more-details.component.html (hunk 10 above)
Comment (leoleader) [reply angular-client/src/components/more-details/more-details.component.html:1]: i do not remember why i put that it is def not needed

File: angular-client/src/components/history/carousel.component.html (hunk 1 above)
Comment (leoleader) [reply angular-client/src/components/history/carousel.component.html:2]: in the docs for the carousel component it uses these ng-templates for its header, content, footer etc. This one specifically allows a run to be displayed as an item of the carousel

**Pull request merged**